File: algorithms/mc/SVP/private/reorth.f
      subroutine reorth(n,k,V,ldv,vnew,normv,index,alpha,work,
     c     iflag,nre)
c
c FORTRAN 77 version of MATLAB routine REORTH:
c
c REORTH   Reorthogonalize a vector using iterated Gram-Schmidt
c
c   [R_NEW,NORMR_NEW,NRE] = reorth(Q,R,NORMR,INDEX,ALPHA,METHOD)
c   reorthogonalizes R against the subset of columns of Q given by INDEX. 
c   If INDEX==[] then R is reorthogonalized all columns of Q.
c   If the result R_NEW has a small norm, i.e. if norm(R_NEW) < ALPHA*NORMR,
c   then a second reorthogonalization is performed. If the norm of R_NEW
c   is once more decreased by  more than a factor of ALPHA then R is 
c   numerically in span(Q(:,INDEX)) and a zero-vector is returned for R_NEW.
c
c   If method==0 then iterated modified Gram-Schmidt is used.
c   If method==1 then iterated classical Gram-Schmidt is used.
c
c   The default value for ALPHA is 0.5. 
c   NRE is the number of reorthogonalizations performed (1 or 2).

c References: 
c  Aake Bjorck, "Numerical Methods for Least Squares Problems",
c  SIAM, Philadelphia, 1996, pp. 68-69.
c
c  J.~W. Daniel, W.~B. Gragg, L. Kaufman and G.~W. Stewart, 
c  ``Reorthogonalization and Stable Algorithms Updating the
c  Gram-Schmidt QR Factorization'', Math. Comp.,  30 (1976), no.
c  136, pp. 772-795.
c
c  B. N. Parlett, ``The Symmetric Eigenvalue Problem'', 
c  Prentice-Hall, Englewood Cliffs, NJ, 1980. pp. 105-109

c  Rasmus Munk Larsen, DAIMI, 1998.
      implicit none
      integer n,k,ldv,i,iflag,nre
      double precision V(ldv,*),vnew(*),normv,index(*),work(*)
      double precision alpha,normv_old,dnrm2
      integer MAXTRY
      parameter (MAXTRY=4)
      external dgemv,dnrm2

c     Hack: If index .ne. 1:k we do MGS to avoid reshuffling.
      if (iflag.eq.1) then     
         do i=1,k
            if (int(index(i)).ne.i) then
               iflag=0
               goto 100
            endif
         enddo
      endif
 100  normv_old = 0
      nre = 0
      normv = dnrm2(n,vnew,1)
      do while ((normv.lt.alpha*normv_old .or. nre.eq.0))
         if (iflag.eq.1) then
c           CGS:
            call dgemv('T',n,k,1D0,V,ldv,vnew,1,0D0,work,1)
            call dgemv('N',n,k,-1D0,V,ldv,work,1,1D0,vnew,1)
         else
c           MGS:
            call MGS(n,k,V,ldv,vnew,index)
         endif
         normv_old = normv
         normv = dnrm2(n,vnew,1)         
         nre = nre + 1
         
         if  ( nre.gt.MAXTRY )  then
c
c     vnew is numerically in span(V) => return vnew = (0,0,...,0)^T
            normv = 0d0
            do i=1,n
               vnew(i) = 0d0
            enddo
            return
         endif
      enddo
      end
c
c****************************************************************************
c

      subroutine MGS(n,k,V,ldv,vnew,index)
      implicit none
      integer n,k,ldv
      double precision V(ldv,*),vnew(*),index(*)
      integer i,j,idx
      double precision s

c     
c     Modified Gram-Schmidt orthogonalization:
c     Orthogalizes vnew against the k vectors in V by the
c     iterative process     
c     
c     FOR i=1...k DO          
c       vnew = vnew - DOT( V(:,i), vnew ) * V(:,i) 
c

c     This simple version is faster on Pentium machines.
c     Compile with "g77 -O6 -funroll-all-loops -fomit-frame-pointer"
     
      do i=1,k
         idx = int(index(i))
         s = 0
         do j=1,n
            s = s + V(j,idx)*vnew(j)
         enddo
         do j=1,n
            vnew(j) = vnew(j) - s*V(j,idx)
         enddo
      enddo
      end
c
c****************************************************************************
c

File: algorithms/st/ReProCS/PROPACK/reorth.m
function [r,normr,nre,s] = reorth(Q,r,normr,index,alpha,method)

%REORTH   Reorthogonalize a vector using iterated Gram-Schmidt
%
%   [R_NEW,NORMR_NEW,NRE] = reorth(Q,R,NORMR,INDEX,ALPHA,METHOD)
%   reorthogonalizes R against the subset of columns of Q given by INDEX. 
%   If INDEX==[] then R is reorthogonalized all columns of Q.
%   If the result R_NEW has a small norm, i.e. if norm(R_NEW) < ALPHA*NORMR,
%   then a second reorthogonalization is performed. If the norm of R_NEW
%   is once more decreased by  more than a factor of ALPHA then R is 
%   numerically in span(Q(:,INDEX)) and a zero-vector is returned for R_NEW.
%
%   If method==0 then iterated modified Gram-Schmidt is used.
%   If method==1 then iterated classical Gram-Schmidt is used.
%
%   The default value for ALPHA is 0.5. 
%   NRE is the number of reorthogonalizations performed (1 or 2).

% References: 
%  Aake Bjorck, "Numerical Methods for Least Squares Problems",
%  SIAM, Philadelphia, 1996, pp. 68-69.
%
%  J.~W. Daniel, W.~B. Gragg, L. Kaufman and G.~W. Stewart, 
%  ``Reorthogonalization and Stable Algorithms Updating the
%  Gram-Schmidt QR Factorization'', Math. Comp.,  30 (1976), no.
%  136, pp. 772-795.
%
%  B. N. Parlett, ``The Symmetric Eigenvalue Problem'', 
%  Prentice-Hall, Englewood Cliffs, NJ, 1980. pp. 105-109

%  Rasmus Munk Larsen, DAIMI, 1998.

% Check input arguments.
% warning('PROPACK:NotUsingMex','Using slow matlab code for reorth.')
if nargin<2
  error('Not enough input arguments.')
end
[n k1] = size(Q);
if nargin<3 | isempty(normr)
%  normr = norm(r);
  normr = sqrt(r'*r);
end
if nargin<4 | isempty(index)
  k=k1;
  index = [1:k]';
  simple = 1;
else
  k = length(index);
  if k==k1 & index(:)==[1:k]'
    simple = 1;
  else
    simple = 0;
  end
end
if nargin<5 | isempty(alpha)
  alpha=0.5; % This choice garanties that 
             % || Q^T*r_new - e_{k+1} ||_2 <= 2*eps*||r_new||_2,
             % cf. Kahans ``twice is enough'' statement proved in 
             % Parletts book.
end
if nargin<6 | isempty(method)
   method = 0;
end
if k==0 | n==0
  return
end
if nargout>3
  s = zeros(k,1);
end


normr_old = 0;
nre = 0;
while normr < alpha*normr_old | nre==0
  if method==1
    if simple
      t = Q'*r;
      r = r - Q*t;
    else
      t = Q(:,index)'*r;
      r = r - Q(:,index)*t;
    end
  else    
    for i=index, 
      t = Q(:,i)'*r; 
      r = r - Q(:,i)*t;
    end
  end
  if nargout>3
    s = s + t;
  end
  normr_old = normr;
%  normr = norm(r);
  normr = sqrt(r'*r);
  nre = nre + 1;
  if nre > 4
    % r is in span(Q) to full accuracy => accept r = 0 as the new vector.
    r = zeros(n,1);
    normr = 0;
    return
  end
end
